### ISSUE TITLE
In .net core 3.0, Json serialization is not consistent when using AddNewtonsoftJson and causes deserialization issue related to Dictionary prop

### ISSUE THREAD

**Issue by ts46235** [aspnet/AspNetCore#16618]
Setup

Newtonsoft.Json: 12.0.2
Microsoft.AspNetCore.Mvc.NewtonsoftJson: 3.0.0

I have testing both from an nunit test project (targeting both 2.2 and 3.0) that does `AddNewtonsoftJson` when in the 3.0 version as well as running the application and hitting with postman

.Net Core 3.0 application setup:
```
services
                .AddMvc(mvcOptions => mvcOptions.EnableEndpointRouting = false)
                .SetCompatibilityVersion(CompatibilityVersion.Version_2_2)
                .AddNewtonsoftJson();
```

If i have this problemDetails object:
```
// problemDetails has 2 extension items
{Microsoft.AspNetCore.Mvc.ValidationProblemDetails}
    Detail: "Please refer to the 'errors' property for additional details."
    Errors: Count = 1
    Extensions: Count = 2
    Instance: "/api/test/complex-binding-from-query"
    Status: 400
    Title: "One or more validation errors occurred."
    Type: "https://tools.ietf.org/html/rfc7231#section-6.5.1"
```
In 2.2, executing `Newtonsoft.Json.JsonConvert.SerializeObject(problemDetails)` returns 
```
{"errors":{"Int":["The value 'a' is not valid for Int."]},"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"One or more validation errors occurred.","status":400,"detail":"Please refer to the 'errors' property for additional details.","instance":"/api/test/complex-binding-from-query","traceId":"0HLQRT7GRPQBS","correlationId":"0HLQRT7GRPQBS"}
```
In 3.0 it returns:
```
{"Errors":{"Int":["The value 'a' is not valid for Int."]},"Type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","Title":"One or more validation errors occurred.","Status":400,"Detail":"Please refer to the 'errors' property for additional details.","Instance":"/api/test/complex-binding-from-query","Extensions":{"traceId":"0HLQRSUICF8QA","correlationId":"0HLQRSUICF8QA"}}
```
The 3.0 version serialized string includes the IDictionary<string,string> property name, `Extensions`, and we can correctly deserialize that string in 3.0. You can see that this property name is left out of the 2.x version.

Issues:

1. This serialized schema is different between these two versions and my clients will have a problem correctly deserializing them. 
2. Using a result filter to return the object in 3.0 does not serialize using Newtonsoft (I assume) and thus is missing the `Extensions` property NAME as seen below:

```
public sealed class ProblemDetailsResultFilterAttribute : Attribute, IAlwaysRunResultFilter
{
        public void OnResultExecuting(ResultExecutingContext context)
        {
             context.Result = new BadRequestObjectResult(problemDetails);
        }
}
```
The response content that gets returned is similar to the 2.2 version...the `Extensions` property name is excluded, but two more differences appear:
1. The key of the item in `Errors` is `param` instead of `Int`
2. The value is missing the ` for Int` suffix).

Deserializing this string causes the `Extensions` property to deserialize as an empty collection (using `Newtonsoft.Json.JsonConvert.DeserializeObject<ValidationProblemDetails>()`)

```
response.Content.ReadAsStringAsync().Result
{"errors":{"param":["The value 'a' is not valid."]},"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"One or more validation errors occurred.","status":400,"detail":"Please refer to the 'errors' property for additional details.","instance":"/api/test/bad-request","traceId":"0HLQRTCDTQ6IK","correlationId":"0HLQRTCDTQ6IK"}
```

Somehow this serialization is not using the same serialization as the Newtonsoft library that we are trying to deserialize with even though we are using `.AddNewtonsoftJson()` and so that is causing our problems. 

Thanks

**Comment by pranavkm** [comment]
@ts46235 `ProblemDetails` and `ValidationProblemDetails` require specific converters to serialize or deserialize correctly. Json.NET's options are configured with these options by default, so that's likely the easiest way to process these:

```C#
public class MyFeature
{
    private readonly MvcNewtonsoftJsonOptions jsonOptions;
    public MyFeature(IOptions<MvcNewtonsoftJsonOptions> jsonOptions) => this.jsonOptions = jsonOptions;
    
    public async Task ReadResponse()
    {
         ....
         JsonConvert.DeserializeObject(problemDetailsString, jsonOptions.SerializerSettings);
     }
}
```

Other options include using the factory to get the serializer settings (https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.NewtonsoftJson/src/JsonSerializerSettingsProvider.cs#L31-L53) or newing up a custom one and adding these two converters:

* https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.NewtonsoftJson/src/ValidationProblemDetailsConverter.cs
* https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.NewtonsoftJson/src/ProblemDetailsConverter.cs

**Comment by ts46235** [comment]
Hello, thank you for the response. Sorry it took a while to respond as I had some work I had to do.

I do see that using those serializerSettings does deserialize correctly, but on that my question is why did it work without needing those in 2.2?....that would cause a code change for that to work.

Secondly, the actual code that I am testing is NOT using JsonConvert manually and instead is returning the result from the filter and then the client is reading the response's content. I only mentioned JsonConvert because that was the only way to validate what I was seeing....sorry for not making that clear, but that is the actual code that is not working and there is not a way to add converters to these calls as it is done somewhere inside the .net core code that I don't see. 

So to summarize, the code I am concerned with is the serialization/deserialization that happens automatically when we set a filter's `Result` and reading the response's `Content`

`context.Result = new BadRequestObjectResult(problemDetails);`

`await response.Content.ReadAsAsync<ValidationProblemDetails>`

**Comment by pranavkm** [comment]
@ts46235 `ReadAsAsync` uses Json.NET under the covers. In 2.x, `ProblemDetails` was annotated with JSON.NET specific attributes that allowed round-tripping without configuring it. I n 3.0, you have to configure `HttpClient` to use the converters I listed in the previous comment. The gist of it should look something like this:

```C#
await response.Content.ReadAsAsync<ValidationProblemDetails>(
    new[] {new JsonMediaTypeFormatter 
    {
          SerializerSettings = new JsonSerializerSettings 
          { 
              Converters =  { new ValidationProblemDetailsConverter(), new ProblemDetailsConverter() }
          }
    });
```

An alternative would be to read the content as string and parsing it using `JsonSerializer`:

```C#
var content = await response.Content.ReadAsStringAsync();
var problemDetails = JsonSerializer.Deserialize(content);
```

We plan on adding support for System.Text.Json to `HttpClient` for the 5.0 release which should help here.

**Comment by ts46235** [comment]
Ok, in my particular case it seems I am only having a problem with the `ValidationProblemDetails` since that is the only one that is running thru my FilterAttribute so it looks like its just an issue when it gets serialized into its Context's `Result`. Adding a converter for it fixed my problem, but...

I could not find this breaking change specified in the Migration literature or release notes. I would think it should be noted that the converters are required now when using `AddNewtonsoftJson()`

Thank you
